cmd/update-main.go

```diff
@@ -68,90 +68,108 @@ const (
 	minioReleaseURL           = "https://dl.minio.io/server/minio/release/" + runtime.GOOS + "-" + runtime.GOARCH + "/"
 )
 
-func getCurrentReleaseTime(minioVersion, minioBinaryPath string) (releaseTime time.Time, err error) {
-	if releaseTime, err = time.Parse(time.RFC3339, minioVersion); err == nil {
+// updateCheckParams - collection of parameters used to check if the
+// version of the current server process is older than the latest
+// available official release. They enable the testability of various
+// routines in the update checking process.
+type updateCheckParams struct {
+	// version of the current server process
+	version string
+	// path to the server binary
+	serverBinaryPath string
+	// file on the server to check for docker env
+	cgroupFile string
+	// URL of release checksum file (contains version of latest
+	// release)
+	releaseChecksumURL string
+	// current process run mode
+	serverMode string
+}
+
+// getCurrentReleaseTime - returns this process's release time.  If it
+// is official minio version, parsed version is returned else minio
+// binary's mod time is returned.
+func (u *updateCheckParams) getCurrentReleaseTime() (releaseTime time.Time,
```

Comment:
> * As you told WRT updateCheckParams is not performance sensitive. the pointer type is good for methods changing fields of the type. This case is not applicable to the struct as none of the methods are updating fields.

Pointer receivers are **required** for methods changing fields of the struct/type. There is no choice in the matter there. The choice is only in the case of methods not updating the struct/type they are attached to.

So what are the considerations in choosing between pointer/value receiver types for such methods? The [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) say that it makes sense for small structs. It also says that you have to think of it as passing all properties of the struct as arguments to each value-receiver method. To me passing 5 variables to so many methods is excessive. So I find a pointer type to be better here. It is not a performance sensitivity argument - but more of an aesthetic to not be wasteful where it is trivial to do so.

You say having pointer receiver has the issue of passing an uninitialized pointer to a method:

> * As I indicated in the example, pointer type must be initialized to avoid run time panic. This is avoidable if methods are to value type.

However, such programmer mistakes are not easily overcome by simply robotically following a convention - see this code:

``` go
package main

import (
	"fmt"
)

type mytype struct{}

func (t mytype) greet() {
	fmt.Println("spock hand")
}

func doSomething(k mytype) {
	k.greet()
}

func main() {
	var t1 *mytype
	doSomething(*t1)
}
```

We have neatly followed the convention and used value-receiver types. However, this also crashes because the caller sent in an unintialized pointer. **The conclusion is that the programmer has to be careful to avoid such pitfalls (especially in Golang). There is no *holy-cow* convention that will help avoid shooting yourself in the foot.**

Again, in this change, the scope of the struct is just one file (ignoring test files for obvious reasons). **It is quite clear that pointers will be always be initialized in this code.** So I still say that it is a fine choice to use pointer-receiver here.


> * method to pointer type is even accessible by nil pointer. Having any methods to pointer type will lead to panic.
> 
> ``` go
> func (t *mytype) greet() {
> 	fmt.Println("spock hand")
> }
> 
> // (fixed a bug here to make it compile)
> var t1 *mytype
> t1.greet()
> ```

Again, I don't understand this point - you say `"Having any methods to pointer type will lead to panic."` but just gave an example yourself of a pointer-receiver method that will [not panic or crash](https://play.golang.org/p/8NKAuImyP1). Here, it appears that it doesn't matter if the method is pointer/value-receiver, because it makes no use of any properties in the type. **Additionally, all the code in the present change are counter-examples to the statement "pointer-receiver methods will lead to panic".**

> * Choosing pointer or value type for method is by considering performance and/or changing field values.

This is not true - as I said earlier, if the method needs to change fields in the struct it has to be a pointer-receiver type. This is compulsory. There is no choice here.

The choice comes only in the case of methods not changing their type's fields. The main consideration in Go coding guidelines *is performance*! Please take some time to go through the [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) - they in fact say to default to pointer types if in doubt.